### ISSUE TITLE
How to determine failed downloads from preheating?

### ISSUE THREAD

**Issue by AssyriaGameStudio** [kean/Nuke#167]
Hi,
So we're currently using Kingfisher, however we're noticing some performance issues on large collection views... So Nuke library seems like the ideal solution!

The only query I have is in regards to the preheating... We're currently trying to download all the images our app needs on-launch so that by the time the users reaches the screen where they're actually needed; they're already available.

We're keen for the app to work in areas where the user has weak signal; often meaning not all image requests are successful. Kingfisher's ImagePrefetcher caters for this by returning which images succeeded and failed downloading; we then log which ones failed and try again at an appropriate time.

I'm struggling to find anything similar in the Nuke documentation; is there a way to determine which images were and weren't successfully preheated? Also is there a means of specifying a count for the number of times to potentially re-try downloading a given image? Or a means of getting a callback when each image in the preheat list has failed / been downloaded successfully?

Many thanks!

**Comment by kean** [comment]
Hi,

> however we're noticing some performance issues on large collection views

Nuke is definitely [much faster on the main thread](https://github.com/kean/Image-Frameworks-Benchmark) than Kingfisher but I would definitely recommend profiling your app first to make sure what the cause of the performance issues actually is.

There is an `ImagePreheater` type in Nuke which is designed to be used in table and collection views to prefetch images which are expected to appear on the display soon. It's designed to be used on fast internet to make sure the user never sees any loading indicators or empty image views. That's how I personally use it and I imagine this use case is very different from the one that you have.

I haven't extended `ImagePreheater` functionality much from the day it was introduced. There is definitely room for adding more features to it - PRs are welcome.

But in your specific use case I would recommend implementing this by hand. It can be as simple as this:

```swift
func startPrefetchingImages(for urls: [URL]) {
    func prefetchImage(for request: ImageRequest) {
        ImagePipeline.shared.loadImage(with: request) { (response, error) in
            if response == nil {
                // Retry after 10 seconds
                DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(10)) {
                    prefetchImage(for: request)
                }
            }
        }
    }

    for url in urls {
        var request = ImageRequest(url: url)
        request.priority = .low // Lower than regular requests

        prefetchImage(for: request)
    }
}
```

The retry logic can be as complex as you'd like it to be. I personally use [RxNuke](https://github.com/kean/RxNuke) with [smart retries](https://kean.github.io/post/smart-retry).

**Comment by AssyriaGameStudio** [comment]
Hey - I have profiled it, and it's deff kingfisher that's the culprit (I actually started delving in to clean up some kingfisher code and got it running a fair bit smoother for my specific use-case, however having looked at Nuke and the benchmarks; it seemed like the better option long-term was simply to just switch over, particularly due to the reduced download time of the assets using the WebP extension).

Thanks for pointing me in the right direction :)!

**Comment by AssyriaGameStudio** [comment]
Hey Kean; think I'm almost there! Just one issue I'm running into still...

On older devices it's taking quite a while to display a cached image (often 2-5 seconds). I gather this might be related to the decodingQueue?

I'm going to be implementing UICollectionViewDataSourcePrefetching throughout; but are there any other ways I can improve the speed with which it gets / decodes the cached images?

Weirdly on initial launch it's actually a better user experience, as it's pre-loads them in the background. On subsequent launches I've got it checking if they're cached ahead of trying to load them / if not it just waits until the user is entering the view containing the image before it starts loading from the cache (I assumed it'd be a few ms, but looks like I was wrong as the result is annoying delay on older devices.)

Alternatively; is there a way to prioritise things in the decoding que so I can just bump item the user is trying to view at the specific moment to the front of the que?

Many thanks!

**Comment by kean** [comment]
Hi,

>  I gather this might be related to the decodingQueue?

It might be it. But it's more likely that the problem is background decompression (see `ImageDecompressor`), not decoding (`Data` -> `Image`). I suggest taking a look at [ImageTaskMetrics](https://github.com/kean/Nuke#performance-metrics-beta) to debug performance issues.

If you load lots of images ahead of time I would recommend disabling some of the pipeline steps. Here's what you can do using existing APIs:

```swift
var request = ImageRequest(url: url)
request.processor = nil // Don't decompress the image
request.memoryCacheOptions.isWriteAllowed = false // Disable mem cache (optional)
```

The image is still going to be decoded (`Data` -> `Image`, quick) but it's not going to be decompressed (slow). The ["deduplication"](https://github.com/kean/Nuke#request-dedupication) is smart enough so that it won't download the same image twice if you then start a request with the same URL but with memory cache enabled & processor enabled.

> Alternatively; is there a way to prioritise things in the decoding que so I can just bump item the user is trying to view at the specific moment to the front of the que?

I have a couple of ideas in backlog regarding improving prefetching.

First, there needs to be an API to skip decoding. I was thinking adding something semantic like `func prefetchImage(.inDiskCache | .inMemoryCache)` to `ImagePipeline` itself. This needs to be a separate API because if you skip decoding, no image is produced - can't complete with "success" `ImageResponse`.

Second, the decoding & processing queues should respect the request priorities - at the moment only data loading queue does.

**Comment by kean** [comment]
So all the time is spent waiting in a decoding queue. That's weird, it's not normal for decoding to take so long (0.671 seconds). Are you loading WebP images? This might explain why it takes so long. If so, I could create a patch today that would make decoding & processing queues respect requests' priorities - this would fix your issue.

**Comment by AssyriaGameStudio** [comment]
Ok yep; I'm thinking:
1) I make the non-transparent images into JPG's on the server assuming they'll be decoded much quicker from the cache.

2) There's still a lot of images that require alpha (hence using WebP for them rather than png as for alpha they're like 1/10 of the size)... So for those if I use your patch to just prioritise them if they're on screen (alongside implementing UICollectionViewDataSourcePrefetching)... Hopefully that should improve UX a fair bit.

**Comment by kean** [comment]
> 1. I make the non-transparent images into JPG's on the server assuming they'll be decoded much quicker from the cache.

The reason why JPEG "decoding" (`Data` -> `UIImage`) is fast is that `UIImage` doesn't actually decompress the data until it needs to (`ImageDecompressor`  forces decompression in a background, if you don't, it's going to happen on the main thread when you display the image). So for JPEG most of the work happens in `ImageDecompressor`, not `ImageDecoder`. If you switch to JPEG you might end up just moving the work from "decoding" stage to "processing" stage.

I don't have much recent experience with WebP, but my guess is that it eagerly decompresses the data. This is why decoding takes so long.

> - `decoding` - creating `UIImage` from `Data`
> - `decompression` - forcing `UIImage` to render its contents in `ImageDecompressor`, this happens during the "processing" stage

I would wait for a patch and see if it improves the situation.

**Comment by kean** [comment]
Pushed a new version with these changes. It works well but there is some technical debt which I'm going to address in 7.2.1. Hope that helps, please close the issue if it does and create a new one if you find some other issues.

**Comment by AssyriaGameStudio** [comment]
Ok, so it's working perfectly now!

The patch above deff helped... That said, WebP decoding was still really slow on older devices (iPad Air 2 / iPad 4th gen etc), to the point it would often still take a few seconds to load the 10-15 WebP images I had displayed on the given page.

For anyone else that runs into the same issue with WebP:

We're basically just downloading loads of app content on launch which for the most part remains the same once downloaded. The solution for us was to add a layer of our own cacheing whereby once the WebP image is downloaded and decoded for the first time we use UIImagePNGRepresentation to cache it on disk as a png file.

We then used Nuke to simply load the png from local on subsequent re-opens of the app, which decodes orders of magnitude quicker than the WebP image, and even on the old devices makes it work perfectly.

Bit hacky, but it does the job well.

**Comment by kean** [comment]
hey @AssyriaGameStudio, I've addressed tech debt introduced in version 7.1 and 7.2 and also made some improvements in the way operations are schedules, reused, canceled, and how priorities are updated. Now it should be pretty much perfect. If you spot any other performance issues, please let me know.

**Comment by zhihuitang** [comment]
Hi @kean 

Theorically, if `prefetchImage` failed, it will try forever? Is my understanding correct?

```
func startPrefetchingImages(for urls: [URL]) {
    func prefetchImage(for request: ImageRequest) {
        ImagePipeline.shared.loadImage(with: request) { (response, error) in
            if response == nil {
                // Retry after 10 seconds
                DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(10)) {
                    prefetchImage(for: request)
                }
            }
        }
    }

    for url in urls {
        var request = ImageRequest(url: url)
        request.priority = .low // Lower than regular requests

        prefetchImage(for: request)
    }
}
```
